### PR TITLE
add support for Forgejo

### PR DESCRIPTION
**Pull request by earl-warren**
Container images including this PR are available for testing at:

* https://codeberg.org/forgejo-contrib/-/packages/container/woodpecker-forgejo-server/versions
* https://codeberg.org/forgejo-contrib/-/packages/container/woodpecker-forgejo-agent/versions
* https://codeberg.org/forgejo-contrib/-/packages/container/woodpecker-forgejo-cli/versions

---

This is a followup of a [previous PR](https://github.com/woodpecker-ci/woodpecker/pull/1535) that was submitted at an early stage of development. The implementation of this Forgejo driver has matured since and is now substantially different. I hope you will consider its inclusion in Woodpecker and I stand ready to address any review comments you may have.

### CODE REVIEW

**Comment by 6543** [comment]
@earl-warren TLDR: please just update tje docs

**Comment by earl-warren** [comment]
> Gitea's and Forgejo's APIs are the same (and will be the same as it's a soft fork).

This is not true: new API will be added, for instance to support semantic versioning. Forgejo API will be a superset of the Gitea API as long as it is a soft fork. It means that the Gitea ecosystem will work with Forgejo. But not the other way around: there is no commitment from Gitea to be compatible with Forgejo.

Here is another example: Forgejo adds a new webhook type "forgejo". This is not going to be supported by Gitea.

Does that clarify why this pull request is proposed?

**Comment by earl-warren** [comment]
These is value in using additional Forgejo API endpoints to improve the driver and I'm sure you'll agree eventually.

**Comment by 6543** [comment]
@earl-warren asap forgejo **do break** with gitea api  a pull to add forgejo dedicated will be accepted until then it's just another implementation to maintain :/

**Comment by earl-warren** [comment]
I'd rather maintain the current Forgejo driver in Woodpecker than in a Woodpecker fork but it is not for me to decide. It's only marginally problematic on my end though.

**Comment by 6543** [comment]
@earl-warren please point out the value of your implementation over the currently used one?

**Comment by earl-warren** [comment]
I pushed an updated version of the Forgejo driver. I understand you see no value in it as long as the Forgejo server does not introduce a breaking change that would make it unusable with the Gitea driver and I respect that. My intention is to keep you informed, not to convince you. I hope this is fine and I would not be offended if you are not interested. I would however appreciate if you do not characterize my work as being redundant or useless.

**Comment by 6543** [comment]
Looks like it is going to fix  #1505 ...

If I'm allowed I will cherry-pick parts of this pull ...

**Comment by earl-warren** [comment]
That's a lot of negative feedback and I'm not sure what to do with it. It is surprising to read that people seem to be angry at me for letting them know of what I'm doing. I'd like to emphasize, again, that I'm completely fine if people are not interested in what I'm doing or how I'm doing it.

Could you just please refrain from characterizing my work as useless, duplicate, bad engineering etc.?

**Comment by 6543** [comment]
well some of your work is de-facto to duplicate code ...
about "bad engineering", I personaly did not compare it to the existing gitea driver so I cant rate it in any meaningfull way atm
about useless, well the idea to do it in the first place might be, but it's done and now it should be compared.
But as wat i saw is to implement the client without a sdk directly into woodpecker ... I'm rather going with a good, well used sdk, maintained separately if possible (go-sdk of gitea in that case).

tldr: I personally never called your work bad in any way just that it redo what's already done( possible )

**Comment by earl-warren** [comment]
There seems to be a misunderstanding about my intent, let me clarify. 

I'm doing work to improve Woodpecker for my own benefit as well as to address specific challenges in the context of the Forgejo project.

To behave like a good Free Software citizen, I communicate with the upstream project (on GitHub despite how much I dislike using it). I also make an effort to comply with what I understand of the best practice (not using panic() but returning err in functions is one example) and publish my work under a compatible license (non copyleft in this case).

It is then up to the maintainers to choose how to react. The feedback I got has been abundantly clear: not interested. And this is fine. I think that will change in the future but in the meantime there is no need to repeat. Maybe just leave this PR evolve and come back later?

**Comment by 6543** [comment]
->  #1733
